### PR TITLE
Create Core class for graph stuff and Handler class for property handling

### PR DESCRIPTION
**Pull request by bergos**
With this patch SimpleRDF
- has a new Core class which contains all the graph read/write stuff
- has a new DefinePropertyHandler class which handles the `Object.defineProperty` implementation
- should be easier to test (hidden functions are now methods)
- class has less `_` properties
- has no longer functions with self parameter (now methods in the right context)
- is more logical structured
- is `Proxy` implementations prepared

### CODE REVIEW

File: lite.js [hunk 1]
```diff
@@ -3,194 +3,74 @@
...
 const SimpleArray = require('./lib/array')
 const SimpleContext = require('./lib/context')
+const SimpleCore = require('./lib/core')
+const SimpleHandler = require('./lib/define-property-handler')
 
 class SimpleRDF {
   constructor (context, iri, graph) {
-    this._iri = buildIri(iri)
-    this._objects = {}
+    this._core = new SimpleCore(this)
```
Comment (nicola) [inline lite.js:11]: This is fine, I am not a big fan of this, but this can work (why not inherit)

**Comment by nicola** [comment]
OK with me, although see comment

File: lite.js (hunk 1 above)
Comment (bergos) [inline lite.js:11]: I don't want to pollute the interface of the SimpleRDF class with all the properties and methods of Core and Handler. I agree that this is usually an anti-pattern, but we have a special use case. SimpleRDF is the base class for everything and we should try to keep everything small :-)

File: test/test.js [hunk 2]
```diff
@@ -240,8 +240,8 @@ describe('simplerdf', () => {
     blog.post = [post]
     post.iri(postIri)
 
-    assert(blog._graph.match(null, 'http://schema.org/post').toArray().shift().object.equals(postIri))
-    assert(blog._graph.match(null, 'http://schema.org/headline').toArray().shift().subject.equals(postIri))
+    assert(blog._core.graph.match(null, 'http://schema.org/post').toArray().shift().object.equals(postIri))
+    assert(blog._core.graph.match(null, 'http://schema.org/headline').toArray().shift().subject.equals(postIri))
```
Comment (nicola) [inline test/test.js:244]: is there a way to access the graph from blog that is not `._core?`

File: lite.js [hunk 3]
```diff
@@ -3,194 +3,74 @@
...
-
-function buildIri (iri) {
-  if (typeof iri === 'string') {
-    return rdf.createNamedNode(iri)
-  } else {
-    return iri || rdf.createBlankNode()
-  }
-}
-
-function updateSubject (graph, oldSubject, newSubject) {
```
Comment (nicola) [inline lite.js:87]: Are we transforming these functions (that are pure),
into a Class?

**Comment by nicola** [comment]
Overall I feel that this is over-refactoring.
I don't like the SimpleCore function being something that we have to create an object for.

For example, if we really want to modularize this, maybe we could have a SimpleCore that is a class with static methods, so that we can avoid the `._core`.

I am not convinced why this is an improvement, it makes the codebase somehow more complex no?

File: test/test.js (hunk 2 above)
Comment (bergos) [inline test/test.js:244]: Yes, with `.graph()`, but I tried to avoid changing the logic of the code.

File: lite.js (hunk 3 above)
Comment (bergos) [inline lite.js:87]: `buildIri` was moved to a static method. `updateSubject` and `updateObject` was always called with `this._graph` and `this._iri`. It's a little bit hidden, but from the logic point of view it was already a method.

**Comment by bergos** [comment]
This PR moves only code, but doesn't add new functionality. The biggest benefit will be for testing. All the functions I moved are hidden until now. If you look at the `addValue*` and `getValue*` functions, they are already very complex, but we can't access them directly. With this PR we can write tests for them (and I will do this in another PR). For example these issue will be much easier to implement with tests:
- [reverse properties](https://github.com/simplerdf/simplerdf/issues/18)
- [datatype support in the context](https://github.com/simplerdf/simplerdf/issues/41)
- [Support for JSON-ld syntax in assignments](https://github.com/simplerdf/simplerdf/issues/38)

Except `buildIri` all functions are in fact methods. Some of them even have a `self` parameter. We started with ES5 and avoided to add `.bind` many times. I think that's the only reason we didn't use real methods before. So if we use static methods, they will all have a parameter just to forward the object instance context. I see this as a bigger anti-pattern.

**Comment by nicola** [comment]
Perfect!

**Pull request merged**